File: Benchmarking/30 Euler/euler13/euler13.jl
#=
Work out the first ten digits of the sum of the following one-hundred 50-digit numbers.
=#
#Answer: 5537376230

#Terminal: Julia euler13.jl data.txt
data = ARGS[1]
function makeData(data)
  f = open(data)
  row = 1
  mat = zeros(Int, 100, 50)
  for l in eachline(f)
    temp = digits(parse(BigInt, l))
    temp2 = reverse(temp)
    col = 1
    for t in temp2
      mat[row, col] = t
      col += 1
    end
    row += 1
  end
  close(f)
  return mat
end

function calc(mat)
  mat_rows, mat_cols = size(mat)
  r = mat_rows
  c = mat_cols
  temp = Int[]

  for col in c:-1:1
    sum = 0
    for row in r:-1:1
      sum += mat[row, col]
    end
    push!(temp, sum)
  end

  i = 1
  while true
    if i < mat_cols
      k = temp[i] % 10
      n = (temp[i] - k) / 10
      temp[i] = k
      temp[i + 1] = temp[i + 1] + n
    else
      k = temp[i] % 10
      n = (temp[i] - k) / 10
      temp[i] = k
      push!(temp, n)
    end
    if temp[end] < 10
      break
    end
    i += 1
  end
  result = ""
  for i=0:9
    result = string(result, temp[end-i])
  end
  return result
end

println(calc(makeData(data)))

File: Benchmarking/30 Euler/euler13/euler13.py
import sys

data = sys.argv[1]
def makeData(data):
	row = 0
	mat = [ [ 0 for i in range(50) ] for j in range(100) ]
	with open(data, "r") as f:
		for line in f:
			temp = []
			line = "".join(line.split())
			for ch in line:
				if ch.isdigit():
					temp.append(int(ch))
			col = 0
			for t in temp:
				mat[row][col] = t
				col += 1
			row += 1
	return mat

def calc(mat):
	r = len(mat)
	c = len(mat[0])
	temp = []
	for col in range(c-1, -1, -1):
		s = 0
		for row in range(r-1, -1, -1):
			s += mat[row][col]
		temp.append(s)
	i = 0
	while True:
		if i < c-1:
			k = temp[i] % 10
			n = (temp[i] - k) / 10
			temp[i] = k
			temp[i + 1] = temp[i + 1] + n
		else:
			k = temp[i] % 10
			n = (temp[i] - k) / 10
			temp[i] = k
			temp.append(n)
		if temp[-1] < 10:
			break
		i += 1
	result = ""
	for i in range(0, 10):
		result = result + str(temp[-(1+i)])
	return result

print calc(makeData(data))
